File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        int i = 0;
        while(i < s.length()){
            if (s.charAt(i) == '{' || s.charAt(i) == '(' || s.charAt(i) == '['){
            stack.add(s.charAt(i));
        }
        else{
            if (stack.isEmpty()){
                return false;
            }
            if (s.charAt(i) == ')' && stack.peek() != '('){
                return false;
            }
            if (s.charAt(i) == ']' && stack.peek() != '['){
                return false;
            }
            if (s.charAt(i) == '}' && stack.peek() != '{'){
                return false;
            }
            stack.pop();
        }
        i++;

    }
        return stack.isEmpty();
}}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        l = ['(', '{', '[']
        openingset = set(l)
        for char in s:
            if char in openingset:
                stack.append(char)
            elif(not stack):
                return False
            elif char == '}':
                temp = stack.pop()
                if (temp != '{'):
                    return False
            elif char == ')':
                temp = stack.pop()
                if (temp != '('):
                    return False                
            elif char == ']':
                temp = stack.pop()
                if (temp != '['):
                    return False        
        return len(stack) == 0
                

